python/torch_mlir/__init__.py

```diff
@@ -13,11 +12,28 @@
 from torch._functorch.compile_utils import strip_overloads
 import torch
 
-from torch_mlir.passmanager import PassManager
+from .dynamo import _get_decomposition_table
 from .compiler_utils import run_pipeline_with_repro_report
 from torch_mlir.dialects.torch.importer.jit_ir import ClassAnnotator, ImportOptions, ModuleBuilder
 from torch_mlir.dialects.torch.importer.jit_ir.build_tools.library_generator import generate_library
 
+import torch.fx
+from torch._dynamo.backends.common import aot_autograd
+from torch._functorch.aot_autograd import make_boxed_compiler, set_model_name, get_aot_compilation_context
+from torch_mlir._dynamo_fx_importer import import_fx_graph_as_func
+
+mlir_module = None
```

Comment:
we should not be using a global here. The right solution uses dynamo.export or similar API's. We can also expose helpers in torch_mlir.dynamo for helping users to plug in their MLIR backends

Reply:
torch_mlir.compile is fundamentally an "AoT" compilation path. If we want an online/JIT Dynamo/torch.compile backend then we will want to expose that as a different API.

Reply:
>we should not be using a global here. The right solution uses dynamo.export or similar API's. We can also expose helpers in torch_mlir.dynamo for helping users to plug in their MLIR backends

I don't understand what this means; `dynamo.export`:

> Export an input function f to a format that can be executed outside of PyTorch using the FX graph.

It uses a trivial compiler `dynamo_normalization_capturing_compiler` to capture the FX graph; there is no support for plugging in your compiler; furthermore `dynamo_normalization_capturing_compiler` is **using nonlocals**:

```python
    def dynamo_normalization_capturing_compiler(
        gm: torch.fx.GraphModule, example_inputs
    ):
        nonlocal graph
        assert (
            graph is None
        ), "Tried to emit a second graph during export. Tracing through 'f' must produce a single graph."
        graph = gm

        def result_capturing_wrapper(*graph_inputs):
            nonlocal graph_captured_result
            nonlocal graph_captured_input

            graph_captured_input = graph_inputs
            assert graph is not None
            graph_captured_result = graph(*graph_inputs)
            return graph_captured_result

        return result_capturing_wrapper
```

So I can change to `nonlocal` but otherwise, I don't see a way.

Reply:
We don't plug in any compiler in the dynamo.export path. It gives us an FX graph (I don't care how it is implemented internally), and we import that. (we might need to file feature requests on their end for functionalization/decomposition and stuff).

I think the best path for the initial integration with dynamo/2.0 is via the JIT mode, i.e. not through torch_mlir.compile.